numpy/core/include/numpy/ndarraytypes.h

```diff
@@ -652,6 +650,27 @@ typedef struct _arr_descr {
         PyObject *shape;       /* a tuple */
 } PyArray_ArrayDescr;
 
+/*
+ * Memory handler structure for array data.
+ */
+typedef void *(PyDataMem_AllocFunc)(size_t size);
+typedef void *(PyDataMem_ZeroedAllocFunc)(size_t nelems, size_t elsize);
+typedef void (PyDataMem_FreeFunc)(void *ptr, size_t size);
+typedef void *(PyDataMem_ReallocFunc)(void *ptr, size_t size);
+typedef void *(PyDataMem_CopyFunc)(void *dst, const void *src, size_t size);
+
+typedef struct {
+    char name[128];  /* multiple of 64 to keep the struct unaligned */
+    PyDataMem_AllocFunc *alloc;
+    PyDataMem_ZeroedAllocFunc *zeroed_alloc;
+    PyDataMem_FreeFunc *free;
+    PyDataMem_ReallocFunc *realloc;
+    PyDataMem_CopyFunc *host2obj;  /* copy from the host python */
+    PyDataMem_CopyFunc *obj2host;  /* copy to the host python */
+    PyDataMem_CopyFunc *obj2obj;  /* copy between two objects */
+} PyDataMem_Handler;
+
```

Comment:
The `PyDataMem_CopyFunc`s are currently unused.

Reply:
Can you explain how you imagine them being used, or should this just be a bunch of `void *reserved`?  `host2obj` and `obj2host` sound like GPU things, but right now we are limited to CPU, since there is a lot of copy code that is DType specific?

Reply:
I guess the would be faster `memcpy` (but compatible)? (In the sense that `memcpy` is always a valid fallback?).  Otherwise, we do not inherit them if we do a view (we could for a typical `view` I guess, right now the code does does not special case view vs. owndata though).  Also, if we export a buffer the information is definitely lost?

If you are aiming for a different device, then things like `np.may_share_memory` may need some thought?